.teamcity/settings.kts

```diff
@@ -952,14 +951,88 @@ object RunCalypsoE2eDesktopTests : BuildType({
 				export BROWSERLOCALE="en"
 				export NODE_CONFIG="{\"calypsoBaseURL\":\"${'$'}{URL%/}\"}"
-				export TEST_FILES=${'$'}(join ',' ${'$'}(ls -1 specs*/**/*spec.js))
+				# Find all spec files but exclude playwright tests.
+				export TEST_FILES=${'$'}(join ',' ${'$'}(find specs*/**/*spec.js -type f -not -path specs-playwright/*))
 
 				yarn magellan --config=magellan.json --max_workers=%E2E_WORKERS% --suiteTag=parallel --local_browser=chrome --mocha_args="--reporter mocha-teamcity-reporter" --test=${'$'}{TEST_FILES}
 			""".trimIndent()
 			dockerImagePlatform = ScriptBuildStep.ImagePlatform.Linux
 			dockerImage = "%docker_image_e2e%"
 			dockerRunParameters = "-u %env.UID% --security-opt seccomp=.teamcity/docker-seccomp.json --shm-size=8gb"
 		}
+		script {
+			name = "Run e2e tests using Playwright (desktop)"
+			scriptContent = """
+				#!/bin/bash
+
+				# Update node
+				. "${'$'}NVM_DIR/nvm.sh" --no-use
+				nvm install
+
+				set -o errexit
+				set -o nounset
+				set -o pipefail
+				shopt -s globstar
+				set -x
+
+				cd test/e2e
+				mkdir temp-playwright
+
+				export LIVEBRANCHES=true
+				export NODE_CONFIG_ENV=test
+				export TEST_VIDEO=true
+				export PLAYWRIGHT_BROWSERS_PATH=0
+
+				# Instructs Magellan to not hide the output from individual `mocha` processes. This is required for
+				# mocha-teamcity-reporter to work.
+				export MAGELLANDEBUG=true
+
+				function join() {
+					local IFS=${'$'}1
+					shift
+					echo "${'$'}*"
+				}
+
+				IMAGE_URL="https://calypso.live?image=registry.a8c.com/calypso/app:build-${BuildDockerImage.depParamRefs.buildNumber}";
+				MAX_LOOP=10
+				COUNTER=0
+
+				# Transform an URL like https://calypso.live?image=... into https://<container>.calypso.live
+				while [[ ${'$'}COUNTER -le ${'$'}MAX_LOOP ]]; do
+					COUNTER=${'$'}((COUNTER+1))
+					REDIRECT=${'$'}(curl --output /dev/null --silent --show-error  --write-out "%{http_code} %{redirect_url}" "${'$'}{IMAGE_URL}")
+					read HTTP_STATUS URL <<< "${'$'}{REDIRECT}"
+
+					# 202 means the image is being downloaded, retry in a few seconds
+					if [[ "${'$'}{HTTP_STATUS}" -eq "202" ]]; then
+						sleep 5
+						continue
+					fi
+
+					break
+				done
+
+				if [[ -z "${'$'}URL" ]]; then
+					echo "Can't redirect to ${'$'}{IMAGE_URL}" >&2
+					echo "Curl response: ${'$'}{REDIRECT}" >&2
+					exit 1
+				fi
+
+				# Decrypt config
+				openssl aes-256-cbc -md sha1 -d -in ./config/encrypted.enc -out ./config/local-test.json -k "%CONFIG_E2E_ENCRYPTION_KEY%"
+
+				# Run the test
+				export BROWSERSIZE="desktop"
+				export BROWSERLOCALE="en"
+				export NODE_CONFIG="{\"calypsoBaseURL\":\"${'$'}{URL%/}\"}"
+				export TEST_FILES=${'$'}(join ',' ${'$'}(ls -1 specs-playwright/**/*spec.js))
+
+				xvfb-run yarn magellan --config=magellan-playwright.json --max_workers=%E2E_WORKERS% --suiteTag=parallel --local_browser=chrome --mocha_args="--reporter mocha-teamcity-reporter" --test=${'$'}{TEST_FILES}
```

Comment:
At this moment the additional TeamCity step to run Playwright is a near-copy of the Selenium based test step, but with one important addition of `xvfb-run` prefixing the actual test call.

Without this, Playwright complains that the xvfb display is inaccessible.

Reply:
Given that it appears TeamCity skips rest of the steps if there is a failure in the preceding step, I may have to reconsider the splitting-up of Selenium and Playwright tests into discrete steps.

Reply:
`xvfb` is started by `test/e2e/lib/mocha-hooks.js`. I'm not 100% sure, but if `playwright` uses a different frame buffer, we won't get screenshots at the end of the test. 

Maybe you need to pass the display number to `playwright` browsers? For WebDriver we do it via `--display=:${global.displayNum}` in `test/e2e/lib/driver-manager.js`

Reply:
I chose to use the Playwright recommended method of `xvfb-run` for a couple of reasons:
1. despite `xvfb` being started in the mocha hooks, with the built-in screenshot and video capture utilities, I no longer see a need to port our existing functions to do both of the above. It is a lot of code that can be removed this way.
2. I also just wanted the demonstration test to work without too much effort :P

Reply:
Sorry, I posted the previous comment a bit prematurely.

The built-in functions that Playwright offers are quite good in that they guarantee the capture of screenshot/video. No faffing with the frame buffer needed at all. 

My next PR in the pipeline deals with the rewrite of the screenshot capture, testing it out on TeamCity at the moment.

> What about moving Playwright tests to its own project? That way we can chose to not run them automatically on every PR until they are ready, they won't delay the "main" e2e build and you'll get faster feedback because you won't wait for all WebDriver e2e tests to complete before running Playwright.

This is what I originally intended when I requested the new project under Calypso in the System Requests P2. I assume I would need admin privilege to create a project from the UI as I have tried adding a new Project in the configuration file and TeamCity simply ignored it.

Reply:
Re: TeamCity.

Sorry, I use the wrong term. I suggest moving it to a separate `buildType` inside the existing `project`. I can create an empty one for Playwright from the UI, then you can add `steps` as necessary to run Playwright. Does that sound good?

Reply:
> Re: TeamCity.
> 
> Sorry, I use the wrong term. I suggest moving it to a separate `buildType` inside the existing `project`. I can create an empty one for Playwright from the UI, then you can add `steps` as necessary to run Playwright. Does that sound good?

Yes, that would be fantastic - thank you!

Reply:
#51172

Reply:
Marking this as resolved as per the comments above. Thanks!